File: PowerSet.java
// 78.

//time - O(2^n)
//space - O(n) for the call stack

class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        
        //edge
        if(nums == null || nums.length == 0)
        {
            return result;
        }
        
        List<Integer> path = new ArrayList<>();
        helper(nums, 0, path, result);
        return result;
    }
    
    private void helper(int[] nums, int index, List<Integer> path, List<List<Integer>> result) {
        
        //path is [] when helper if first called, [] is part of output, so add to result
        //every [] at the root of a subtree is part of result
        result.add(new ArrayList<>(path));
        
        //base
        if(index == nums.length)
        {
            return;
        }
        
        for(int i = index; i < nums.length; i++)
        {
            //add current to path
            path.add(nums[i]);
            //recurse
            helper(nums, i + 1, path, result);
            //bactrack
            path.remove(path.size() - 1);
        }
    }
}

File: powerset.py

#Time Complexity :  O(exponential))
#Space Complexity : O(N)
#Did this code successfully run on Leetcode : yes
#Any problem you faced while coding this : no

class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        
        if len(nums)==0:
            return []
        result = []
        n = len(nums)
 
        def helper(index,path):
            if index ==n:
                result.append(list(path))
                return
                
            #dont choose
            helper(index+1,path)
            #choose
            path.append(nums[index])
            helper(index+1,path)
            path.pop()
            
            
        helper(0,[])
        
        return result